tests/integration/ha/test_replication.py

```diff
@@ -0,0 +1,138 @@
+    second_kafka_name = f"{APP_NAME}-two"
+    second_zk_name = f"{ZK_NAME}-two"
+
+    await asyncio.gather(
+        ops_test.model.deploy(
+            kafka_charm, application_name=second_kafka_name, num_units=1, series="jammy"
+        ),
+        ops_test.model.deploy(
+            ZK_NAME, channel="edge", application_name=second_zk_name, num_units=1, series="jammy"
+        ),
+    )
+
+    await ops_test.model.wait_for_idle(
+        apps=[second_kafka_name, second_zk_name],
+        idle_period=30,
+        timeout=3600,
+    )
+    assert ops_test.model.applications[second_kafka_name].status == "blocked"
+
+    await ops_test.model.add_relation(second_kafka_name, second_zk_name)
+    await ops_test.model.wait_for_idle(
+        apps=[second_kafka_name, second_zk_name, APP_NAME],
+        idle_period=30,
+    )
+
+    produce_and_check_logs(
+        model_full_name=ops_test.model_full_name,
+        kafka_unit_name=f"{APP_NAME}/0",
+        provider_unit_name=f"{DUMMY_NAME}/0",
+        topic="hot-topic",
+    )
+
+    # Check that logs are not found on the second cluster
+    with pytest.raises(AssertionError):
+        check_logs(
+            model_full_name=ops_test.model_full_name,
+            kafka_unit_name=f"{second_kafka_name}/0",
+            topic="hot-topic",
+        )
+
+
+async def test_replicated_events(ops_test: OpsTest):
+    await ops_test.model.applications[APP_NAME].add_units(count=2)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", timeout=600, idle_period=120, wait_for_exact_units=3
+    )
+    logger.info("Producing messages and checking on all units")
+    produce_and_check_logs(
+        model_full_name=ops_test.model_full_name,
+        kafka_unit_name=f"{APP_NAME}/0",
+        provider_unit_name=f"{DUMMY_NAME}/0",
+        topic="replicated-topic",
+        replication_factor=3,
+        num_partitions=1,
+    )
+    # check logs in the two remaining units
+    check_logs(
+        model_full_name=ops_test.model_full_name,
+        kafka_unit_name=f"{APP_NAME}/1",
+        topic="replicated-topic",
+    )
+    check_logs(
+        model_full_name=ops_test.model_full_name,
+        kafka_unit_name=f"{APP_NAME}/2",
+        topic="replicated-topic",
+    )
+
+
+async def test_remove_topic_leader(ops_test: OpsTest):
+    leader_num = await get_topic_leader(ops_test=ops_test, topic="replicated-topic")
+    logger.info(f"Killing broker of leader for topic 'replicated-topic': {leader_num}")
+    await kill_unit_process(
+        ops_test=ops_test, unit_name=f"{APP_NAME}/{leader_num}", kill_code="SIGKILL"
+    )
+    # Give time for the service to restart
+    time.sleep(15)
+    # Check that is still possible to write to the same topic.
+    produce_and_check_logs(
+        model_full_name=ops_test.model_full_name,
+        kafka_unit_name=f"{APP_NAME}/0",
+        provider_unit_name=f"{DUMMY_NAME}/0",
+        topic="replicated-topic",
+        create_topic=False,
+    )
```

Comment:
Yes, I agree. Killing a unit and after it comes back to the cluster making sure we can write to the topic is not HA. We should allow for continuous write while the unit goes down. The client of course should handle client disconnections of one broker, but the cluster should be able to function with still 2 out of 3 units, otherwise we are not HA

Reply:
BTW...he we should also check that the partition `leader` has changed

Reply:
Issue right now with `kafka-test-app` is that it needs to be modified. At the moment it only writes a set amount of messages to the cluster, and that is not suitable in general for these tests. I need to update it with a more flexible approach _or_ change the internal app to allow for a continuous amount of messages until stop signal is issued.